pmdastatsd new QA and small agent changes

Erbenos opened the pull request:
New QA has better coverage than the old one. Includes a Valgrind test.

pmdastatsd received small changes to _double_ metric values handling and improvements to clean up procedures.

lzap commented on qa/statsd/src/cases/04.py:
```diff
@@ -0,0 +1,152 @@
+# - statsd.pmda.dropped (before: 0, after: 40)
+# - statsd.pmda.time_spent_parsing (before: 0, after: non-zero)
+# - statsd.pmda.time_spent_aggregating (before: 0, after: non-zero)
+# - statsd.pmda.metrics_tracked, with its counter, gauge, duration and total instances (before: 0, 0, 0, 0; after: 4, 4, 2, 10)
```
I don't know how these tests work, but is this supposed to check also the before/after state?

lzap commented on qa/statsd/src/cases/09.py:
```diff
@@ -0,0 +1,81 @@
+expected_percentile90 = 18000000
+expected_percentile95 = 19000000
+expected_percentile99 = 19800000
+expected_stddev = 5773500.278068
```
Ah disregard, I see how this is done here.

lzap commented on qa/statsd/src/cases/12.py:
```diff
@@ -0,0 +1,157 @@
+    "cache_cleared:1|msa",
+    "session_started:|ms",
+    ":20|ms"
+]
```
Nitpick: If these "bad" metric names are not needed for these test, I'd probably drop these as these were already tested above.

Erbenos replied:
12.py checks if the specificity of various verbose levels is ordered correctly. Payloads as above would generate additional err logs.

lzap commented on qa/statsd/src/cases/13.py:
```diff
@@ -0,0 +1,76 @@
+
+run_test()
+
+
```
This test does only check of there are any errors I assume? Would it be feasible also to check those metrics created?

Erbenos replied:
08.py checks if labels are OK, the payload is nearly identical as in case of 13.py

lzap commented:
Tests look good.

natoscott commented on qa/statsd/src/cases/09.py:
```diff
@@ -0,0 +1,81 @@
+            if utils.check_is_in_bounds(expected_average, number_value):
+                status = True
+        elif k == "/count":
+            # TODO: Ask Nathan, if this is OK
```
Which part?  The 0.5?

Erbenos replied:
Yes, that one.

natoscott replied:
Yes, that's fine - it's a common pattern - when writing shell tests we even have a shared _within_tolerance() function available for this purpose.

Erbenos replied:
Alright, I will leave it as is and remove the TODO comment, thanks.

Erbenos replied:
I had to set up a new VM to work on this on which I encounter following phenomena: even 0.5 toleration margin is not enough, particularly in relation to count of received UDP payloads, even after setting ```sysctl net.core.rmem_max``` and ```sysctl net.core.rmem_default``` to 26214400, which makes me worry about reliability of such a test - even though its UDP after all, what would you approach to this be?

natoscott replied:
Set the acceptable boundaries sufficiently wide that any reasonable scenario would be catered for.  Add comments to the test discussing possible causes for values outside these boundaries, in case people have to come along later and widen the boundaries further.  Even with wide boundaries, its often still likely to be a worthwhile test.

Erbenos replied:
Will do, thanks.

natoscott commented on qa/statsd/src/cases/06.py:
```diff
@@ -0,0 +1,83 @@
+        sock.sendto("test_gauge2:+{}|g".format(overflow_payload).encode("utf-8"), (ip, port))
+        utils.print_metric("statsd.pmda.dropped")
+        utils.print_metric("statsd.test_gauge2")
+        # TODO: check if this is truly the desired behavior
```
Now would be a good time to check, and figure out an answer (before this is merged) - otherwise, we're all likely to forget over time.

Erbenos replied:
I actually already did the other day (we had small convo in Slack about it), emailed it trough with Lukas if it is OK and we resolved it. I will remove the comment in code.

natoscott commented on qa/statsd/GNUmakefile:
```diff
@@ -2,14 +2,50 @@ TOPDIR = ../..
 	$(INSTALL) -m 755 -d $(TESTDIR)
-	$(INSTALL) -m 644 $(RBFILES) $(TESTDIR)
+	# I wish I knew how to extract such listing from $(TESTFILES), not sure why this is needed yet
+	$(INSTALL) -m 644 src/cases/01.py $(TESTDIR)/src/cases/01.py
```
In the same way TESTDIR is installed above, each new directory used also needs to be $(INSTALL)'d ... otherwise, the packaging ends up slightly incorrect (nothing 'owns' the directory is its not explicitly installed).

Erbenos replied:
I added GNUmakefile's to each qa/statsd dir in 78b11bf (as inspired by qa/slurm). I guess GNUmakefile.install are for post install steps once all test files are `$(INSTALL)`-ed? Will leave only `default setup install clean check` in them since python is interpreted.

natoscott replied:
> I added GNUmakefile's to each qa/statsd dir in [78b11bf](https://github.com/performancecopilot/pcp/commit/78b11bfbc9987b9f97ed2b788ee88ae891abddce) (as inspired by qa/slurm).

Perfect.

> I guess GNUmakefile.install are for post install steps once all test files are `$(INSTALL)`-ed? 

This is the makefile that will be in the pcp-testsuite (rpm/deb) package, used when tests are run from /var/lib/pcp/testsuite

natoscott commented on qa/statsd/GNUmakefile:
```diff
@@ -2,14 +2,50 @@ TOPDIR = ../..
+	$(INSTALL) -m 644 src/cases/13.py $(TESTDIR)/src/cases/13.py
+	$(INSTALL) -m 644 src/cases/14.py $(TESTDIR)/src/cases/14.py
+	$(INSTALL) -m 644 src/cases/15.py $(TESTDIR)/src/cases/15.py
+	$(INSTALL) -m 644 src/configs/complex/0/pmdastatsd.ini $(TESTDIR)/src/configs/complex/0/pmdastatsd.ini
```
Likewise for each of these new ini file dirs :|

Since these configs are so small, I'd consider generating them at runtime instead of committing them as individual files in the repo.

natoscott commented on qa/1599:
```diff
@@ -15,8 +15,9 @@ echo "QA output created by $seq"
 
 test -e $PCP_PMDAS_DIR/statsd/pmdastatsd || _notrun "statsd PMDA not installed"
 
```
The comment at the start of this file refering to 'ruby' needs updating.

natoscott commented on qa/1599:
```diff
@@ -15,8 +15,9 @@ echo "QA output created by $seq"
 
```
Consider replacing the above three lines with common.python now.

natoscott commented on qa/1599:
```diff
@@ -15,8 +15,9 @@ echo "QA output created by $seq"
 
-# NOTE: Miroslav is planning to re-work this in python/shell
-which ruby >/dev/null 2>&1 || _notrun "ruby not installed"
+which python >/dev/null 2>&1 || _notrun "python not installed"
```
common.python will have checked this (python not installed) for you already.

natoscott commented on qa/1599:
```diff
@@ -28,41 +29,18 @@ status=1	# failure is the default!
 for script in $scripts
 do
-    ruby $script
+    python $script $here/statsd/output
```
Using common.python, this should then be $python rather than just python.

This lets us test both python2 and python3 more easily (we use python3 more aggressively - whenever available - over python2).

Erbenos commented:
Thank you for your review so far, I will address the issues you brought up.

Erbenos commented on qa/1599:
```diff
@@ -15,8 +15,9 @@ echo "QA output created by $seq"
-which ruby >/dev/null 2>&1 || _notrun "ruby not installed"
+which python >/dev/null 2>&1 || _notrun "python not installed"
+
+which valgrind >/dev/null 2>&1 || _notrun "valgrind not installed"
```
Addresed by 78b11bf

Erbenos commented:
The commit above addressed issues mentioned but also disabled 13.py as I noticed the test case is not deterministic and will need to be filtered/processed in python further.
